File: app/src/main/java/com/wang/example/MainActivity.java
package com.wang.example;

import android.content.Intent;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.wang.container.BaseContainerAdapter;
import com.wang.container.interfaces.OnItemClickListener;
import com.wang.example.msg.adapter.ImgAdapter;
import com.wang.example.msg.adapter.PaySuccessOrderAdapter;
import com.wang.example.msg.adapter.TextAdapter;
import com.wang.example.msg.adapter.UnsupportedAdapter;
import com.wang.example.msg.adapter.WaitPayOrderAdapter;
import com.wang.example.msg.bean.BaseMsgBean;
import com.wang.example.msg.bean.TestData;
import com.wang.example.utils.ToastUtils;

import java.util.Collections;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    private RecyclerView mRv;
    @Nullable
    private BaseContainerAdapter mBaseAdapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById(R.id.tv_msg).setOnClickListener(v -> createMsg());
        findViewById(R.id.tv_other).setOnClickListener(v -> {
            ToastUtils.toast("暂时没精力写");
        });
        findViewById(R.id.tv_reverse).setOnClickListener(v -> {
            if (mBaseAdapter != null) {
                Collections.reverse(mBaseAdapter.getList());
                mBaseAdapter.notifyDataSetChanged();
            }
        });
        findViewById(R.id.tv_delete).setOnClickListener(v -> {
            if (mBaseAdapter != null) {
                int random = (int) (Math.random() * (Math.min(mBaseAdapter.size(), 5)));
                if (random < mBaseAdapter.size()) {
                    mBaseAdapter.getList().remove(random);
                    //这里需要注意，header的存在
                    mBaseAdapter.notifyItemRemoved(mBaseAdapter.getHeaderView() == null ? random : random + 1);
                    ToastUtils.toast("你删除了list position：" + random);

                    //remove动画不会改变第一条和最后一条的状态，所以带header、footer提示的话也要同步刷新
                    mBaseAdapter.notifyItemChanged(mBaseAdapter.getHeaderView() == null ? 0 : 1);
                    mBaseAdapter.notifyItemChanged(mBaseAdapter.getHeaderView() == null ? mBaseAdapter.size() - 1 : mBaseAdapter.size());
                }
            }
        });

        mRv = findViewById(R.id.rv_main);
        mRv.setLayoutManager(new LinearLayoutManager(this));
    }

    private void createMsg() {
        final BaseContainerAdapter<BaseMsgBean> baseAdapter = new BaseContainerAdapter<>();
        baseAdapter.addAdapter(new TextAdapter(), new ImgAdapter(), new WaitPayOrderAdapter(),
                new PaySuccessOrderAdapter(), new UnsupportedAdapter());

        //header、footer
        TextView headerView = new TextView(this);
        headerView.setText("这是header");
        headerView.setGravity(Gravity.CENTER);
        headerView.setPadding(50, 50, 50, 50);
        TextView footerView = new TextView(this);
        footerView.setText("这是footer");
        footerView.setGravity(Gravity.CENTER);
        footerView.setPadding(50, 50, 50, 50);
        footerView.setOnClickListener(v -> ToastUtils.toast("你点击了footer"));
        baseAdapter.setHeaderView(headerView);
        baseAdapter.setFooterView(footerView);

        final List<BaseMsgBean> list = TestData.createMsgList();//模拟请求数据
        List<BaseMsgBean> newList = BaseMsgBean.formatListData(list);//格式化数据
        baseAdapter.setListAndNotifyDataSetChanged(newList);

        //和子adapter的事件都调用会触发，注意自己的逻辑别和子adapter重复
        baseAdapter.setOnItemClickListener(new OnItemClickListener<BaseMsgBean>() {
            @Override
            public void onItemClick(@NonNull View view, int position) {
                int absPosition = baseAdapter.getAbsPosition(getCurrentBean(), position);
                ToastUtils.toast("Base的点击事件，绝对位置：" + absPosition);
            }
        });
        mRv.setAdapter(baseAdapter);
        mBaseAdapter = baseAdapter;
    }


    public void startActivity(Class c) {
        startActivity(new Intent(this, c));
    }
}

File: app/src/main/java/com/wang/example/main/MainActivity.kt
package com.wang.example.main

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.wang.container.BaseContainerAdapter
import com.wang.example.R
import com.wang.example.main.adapter.*
import com.wang.example.main.bean.BaseMsgBean
import com.wang.example.main.bean.TestData
import com.wang.example.two.TwoActivity
import com.wang.example.utils.toast

class MainActivity : AppCompatActivity() {
    private var rv: RecyclerView? = null
    private var baseAdapter: BaseContainerAdapter<*>? = null

    @SuppressLint("NotifyDataSetChanged")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        findViewById<View>(R.id.tv_msg).setOnClickListener { createMsg() }
        findViewById<View>(R.id.tv_other).setOnClickListener { startActivity(TwoActivity::class.java) }
        findViewById<View>(R.id.tv_reverse).setOnClickListener {
            baseAdapter?.let {
                it.list.reverse()
                it.notifyDataSetChanged()
            }
        }
        findViewById<View>(R.id.tv_delete).setOnClickListener { _ ->
            baseAdapter?.let { it ->
                val random = (Math.random() * it.listSize().coerceAtMost(5)).toInt()
                if (random < it.listSize()) {
                    it.list.removeAt(random)
                    //这里需要注意，header的存在
                    it.notifyItemRemoved(random + it.headerViewCount)
                    "你删除了list position：$random".toast()

                    //remove动画不会改变第一条和最后一条的状态，所以带header、footer提示的话也要同步刷新
                    it.notifyItemChanged(it.headerViewCount)
                    it.notifyItemChanged(it.listSize() + it.headerViewCount - 1)
                }
            }
        }
        rv = findViewById(R.id.rv_main)
        rv?.layoutManager = LinearLayoutManager(this)
    }

    private fun createMsg() {
        val ba = BaseContainerAdapter<BaseMsgBean>()
        ba.addAdapter(
            TextAdapter(), ImgAdapter(), WaitPayOrderAdapter(),
            PaySuccessOrderAdapter(), UnsupportedAdapter()
        )

        //header、footer
        val headerView = TextView(this)
        headerView.text = "这是header"
        headerView.gravity = Gravity.CENTER
        headerView.setPadding(50, 50, 50, 50)
        ba.headerView = headerView
        ba.setFooterView(this, R.layout.adapter_main_footer)
        ba.footerView?.setOnClickListener { _ -> "你点击了footer".toast() }
        val list = TestData.createMsgList() //模拟请求数据
        val newList = BaseMsgBean.formatListData(list) //格式化数据
        ba.setListAndNotify(newList)

        rv?.adapter = ba
        baseAdapter = ba
    }

    fun startActivity(c: Class<*>) {
        startActivity(Intent(this, c))
    }

    companion object {
        const val TAG = "MainActivity"
    }
}